mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+    }
+    if (isLoadStore) {
+      auto elementType = memref.getElementType();
+      unsigned sizeInBits = 0;
+      if (elementType.isIntOrFloat())
+        sizeInBits = elementType.getIntOrFloatBitWidth();
+  
+      // Store element size to be later used to check whether two
+      // accesses are within a cache-line/elementSize distance apart
+      // for a useful locality.
+      this->elementSize[op] = llvm::divideCeil(sizeInBits, 8);
+    }
+  });
+}
+
+/// Get all permutations of the loop nest rooted at `rootForOp`. Store them in
+/// `validPermutations` as a list.
+void getAllValidPermutations(
+    AffineForOp &rootForop, unsigned noLoopsInNest,
+    SmallVector<SmallVector<unsigned, 4>, 8> &validPermutations) {
+
+  SmallVector<unsigned, 4> permutation;
+  for (unsigned i = 0; i < noLoopsInNest; i++)
+    permutation.push_back(i);
+
+  validPermutations.push_back(permutation);
+  SmallVector<AffineForOp, 4> perfectLoopNest;
+  getPerfectlyNestedLoops(perfectLoopNest, rootForop);
+
+  while (std::next_permutation(permutation.begin(), permutation.end()))
+    if (isValidLoopInterchangePermutation(perfectLoopNest, permutation))
+      validPermutations.push_back(permutation);
+}
+
+/// Checks which loops in the loop nest have loop-carried-dependencies.
+void LoopInterchange::getAffineLoopCarriedDependencies(
+    AffineForOp &rootForOp, ArrayRef<Operation *> loadAndStoreOps,
+    unsigned loopNestSize) {
+
+  unsigned numOps = loadAndStoreOps.size();
+  // Resize the `loopCarriedDependenceVector` to loop nest size.
+  this->loopCarriedDependenceVector.resize(loopNestSize);
+
+  for (unsigned i = 0; i < numOps; ++i) {
+    Operation *srcOp = loadAndStoreOps[i];
+    for (unsigned j = 0; j < numOps; ++j) {
+      Operation *dstOp = loadAndStoreOps[j];
+      for (unsigned depth = 1; depth <= loopNestSize + 1; ++depth) {
+        MemRefAccess srcAccess(srcOp);
+        MemRefAccess dstAccess(dstOp);
+        FlatAffineConstraints dependenceConstraints;
+        SmallVector<DependenceComponent, 2> depComps;
+        DependenceResult result = checkMemrefAccessDependence(
+            srcAccess, dstAccess, depth, &dependenceConstraints, &depComps);
+
+        if (hasDependence(result)) {
+          for (unsigned i = 0; i < depComps.size(); i++) {
+            DependenceComponent depComp = depComps[i];
+            if (depComp.lb.getValue() != 0 || depComp.ub.getValue() != 0)
+              this->loopCarriedDependenceVector[i] = true;
+          }
+          // A dependence is found, hence no need to check on further depths.
+          break;
+        }
+      }
+    }
+  }
+}
+
+/// Calculates a representative cost of a permutation regarding parallelism on
+/// multicores. A permutation having more free outer loops gets a smaller cost.
+uint64_t getParallelismCost(ArrayRef<unsigned> perm,
+                            SmallVector<bool, 4> &loopCarriedDV,
+                            SmallVector<unsigned, 4> &iterCounts) {
```

Comment:
Pass as `ArrayRef` for inputs / read-only arguments.

Reply:
Done.